transport/compress.go

```diff
@@ -0,0 +1,303 @@
+package transport
+
+import (
+	"encoding/binary"
+	"errors"
+	"io"
+
+	"github.com/mmatczuk/scylla-go-driver/frame"
+
+	"github.com/klauspost/compress/s2"
+	"github.com/pierrec/lz4/v4"
+)
+
+const (
+	maxBlockSize = 4 << 20
+	checksumSize = 4
+)
+
+var (
+	ErrCorrupt     = errors.New("s2: corrupt input")
+	ErrUnsupported = errors.New("s2: unsupported input")
+	ErrEnd         = errors.New("end") // This is not technically an error.
+)
+
+type comprReader interface {
+	Read(p []byte) (n int, err error)
+	SetExpectedLen(n int)
+}
+
+type SnappyReader struct {
+	r io.Reader
+	// Maximum block size allowed.
+	maxBlock int
+	// Maximum expected buffer size.
+	maxBufSize int
+	// Initial buffer size.
+	lazyBuf int
+
+	decoded []byte
+	buf     []byte
+
+	// decoded[i:j] contains decoded bytes that have not yet been passed on.
+	i, j int
+
+	err error
+
+	l int
+}
+
+func NewReader(r io.Reader) *SnappyReader {
+	nr := SnappyReader{
+		r:        r,
+		maxBlock: maxBlockSize,
+		lazyBuf:  ioBufferSize,
+	}
+	nr.maxBufSize = s2.MaxEncodedLen(nr.maxBlock) + checksumSize
+	nr.buf = make([]byte, s2.MaxEncodedLen(nr.lazyBuf)+checksumSize)
+	return &nr
+}
+
+func (r *SnappyReader) SetExpectedLen(n int) {
+	r.l = n
+}
+
+func (r *SnappyReader) Read(p []byte) (n int, err error) {
+	if err != nil {
+		return 0, err
+	}
+	for {
+		if r.i < r.j {
+			n := copy(p, r.decoded[r.i:r.j])
+			r.i += n
+			return n, ErrEnd
+		}
+
+		r.buf, r.err = ensureBufferSize(r.l, r.buf, r.maxBufSize)
+		if err != nil {
+			return 0, r.err
+		}
+
+		if r.err = readFull(r.r, r.buf[:r.l]); r.err != nil {
+			return 0, r.err
+		}
+
+		// TODO: Why we need to do this?
+		if zeroes(r.buf[:r.l]) {
```

Comment:
This is a band-aid solution.
Scylla when sending `Ready` response seem to also send body of 5 zeroes.
I am not sure of its cause.